Sources/core/objects/Gun.cpp

```diff
@@ -0,0 +1,164 @@
+}
+
+void Gun::ShootImpl(VDir target, const std::string& sound, const std::string& projectile_type, const std::string& casing_type)
+{
+    id_ptr_on<CubeTile> tile = GetOwner()->GetOwner();
+    id_ptr_on<IMovable> shooter = GetOwner();
+    if (tile.valid())
+    {	
+        if (UseAmmo())
+        {
+            int x = target.x;
+            int y = target.y;
+            Dir facing = shooter->GetDir();
+            if(std::abs(y) != std::abs(x))
+            {
+                shooter->Rotate(VDirToDir(target));
+                auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(VDirToDir(target)));
+                projectile->MakeMovementPattern(target, facing);
+            }
+            else
+            {
+                if (y < 0)
+                {
+                    if (facing == 0|| facing == 1)
+                    {
+                        if (x > 0)
+                        {
+                            shooter->Rotate(D_RIGHT);
+                            auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_RIGHT));
+                            projectile->MakeMovementPattern(target, facing);
+                        }
+                        else
+                        {
+                            shooter->Rotate(D_LEFT);
+                            auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_LEFT));
+                            projectile->MakeMovementPattern(target, facing);
+                        }
+                    }
+                    else
+                    {
+                        shooter->Rotate(D_UP);
+                        auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_UP));
+                        projectile->MakeMovementPattern(target, facing);
+                    }
+                }
+                else
+                {
+                    if (facing == 0|| facing == 1)
+                    {
+                        if (x > 0)
+                        {
+                            shooter->Rotate(D_RIGHT);
+                            auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_RIGHT));
+                            projectile->MakeMovementPattern(target, facing);
+                        }
+                        else
+                        {
+                            shooter->Rotate(D_LEFT);
+                            auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_LEFT));
+                            projectile->MakeMovementPattern(target, facing);
+                        }
+                    }
+                    else
+                    {
+                        shooter->Rotate(D_DOWN);
+                        auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_DOWN));
+                        projectile->MakeMovementPattern(target, facing);
+                    }
```

Comment:
Overall, it is possible to reduce code a lot here.
Instead of
```
shooter->Rotate(D_DOWN);
auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(D_DOWN));
projectile->MakeMovementPattern(target, facing);
```
just create new variable `shooting_dir` and set it in the `if`/`else` branches.
Something like that:
```
else
{
    shooting_dir = D_DOWN
}
```
After that place
```
shooter->Rotate(shooting_dir);
auto projectile = Create<Projectile>(projectile_type, tile->GetNeighbour(shooting_dir));
projectile->MakeMovementPattern(target, facing);
```
once after the end of `if`/`else`, and that should be enough.